### ISSUE TITLE
Dynamic Fine Tuning, an improvement of SFT

### ISSUE THREAD

**Issue by 1485840691** [huggingface/trl#3877]
### Feature request

https://github.com/yongliang-wu/DFT

The change is simple in SFTTrainer

def compute_loss(...):
        if "labels" in inputs and not self.args.use_liger_kernel:
            shift_logits = outputs.logits[..., :-1, :].contiguous()
            shift_labels = inputs["labels"][..., 1:].contiguous()

            **probs = torch.softmax(shift_logits, dim=-1)
            prob_coefficients = probs.gather(1, shift_labels.unsqueeze(-1)).squeeze(-1)
            loss = loss * prob_coefficients.detach()**

### Motivation

Abstract of the paper:
"
We present a simple yet theoretically motivated improvement to Supervised Fine-Tuning (SFT) for the Large Language Model (LLM), addressing its limited generalization compared to reinforcement learning (RL). Through mathematical analysis, we reveal that standard SFT gradients implicitly encode a problematic reward structure that may severely restrict the generalization capabilities of model. To rectify this, we propose Dynamic Fine-Tuning (DFT), stabilizing gradient updates for each token by dynamically rescaling the objective function with the probability of this token. Remarkably, this single-line code change significantly outperforms standard SFT across multiple challenging benchmarks and base models, demonstrating greatly improved generalization. Additionally, our approach shows competitive results in offline RL settings, offering an effective yet simpler alternative. This work bridges theoretical insight and practical solutions, substantially advancing SFT performance.
"

### Your contribution

I will submit PR for it given the code change is small.

**Comment by qgallouedec** [comment]
Closed by #4042
